packages/yew/src/callback.rs

```diff
@@ -9,92 +9,194 @@ use std::rc::Rc;
 
 use crate::html::ImplicitClone;
 
-/// Universal callback wrapper.
-///
-/// An `Rc` wrapper is used to make it cloneable.
-pub struct Callback<IN, OUT = ()> {
-    /// A callback which can be called multiple times
-    pub(crate) cb: Rc<dyn Fn(IN) -> OUT>,
-}
+macro_rules! generate_callback_impls {
+    ($callback:ident, $in_ty:ty, $out_var:ident => $out_val:expr) => {
+        impl<IN, OUT, F: Fn($in_ty) -> OUT + 'static> From<F> for $callback<IN, OUT> {
+            fn from(func: F) -> Self {
+                $callback { cb: Rc::new(func) }
+            }
+        }
 
-impl<IN, OUT, F: Fn(IN) -> OUT + 'static> From<F> for Callback<IN, OUT> {
-    fn from(func: F) -> Self {
-        Callback { cb: Rc::new(func) }
-    }
-}
+        impl<IN, OUT> Clone for $callback<IN, OUT> {
+            fn clone(&self) -> Self {
+                Self {
+                    cb: self.cb.clone(),
+                }
+            }
+        }
 
-impl<IN, OUT> Clone for Callback<IN, OUT> {
-    fn clone(&self) -> Self {
-        Self {
-            cb: self.cb.clone(),
+        #[allow(clippy::vtable_address_comparisons)]
+        impl<IN, OUT> PartialEq for $callback<IN, OUT> {
+            fn eq(&self, other: &$callback<IN, OUT>) -> bool {
+                let ($callback { cb }, $callback { cb: rhs_cb }) = (self, other);
+                Rc::ptr_eq(cb, rhs_cb)
+            }
         }
-    }
-}
 
-#[allow(clippy::vtable_address_comparisons)]
-impl<IN, OUT> PartialEq for Callback<IN, OUT> {
-    fn eq(&self, other: &Callback<IN, OUT>) -> bool {
-        let (Callback { cb }, Callback { cb: rhs_cb }) = (self, other);
-        Rc::ptr_eq(cb, rhs_cb)
-    }
-}
+        impl<IN, OUT> fmt::Debug for $callback<IN, OUT> {
+            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+                write!(f, "$callback<_>")
+            }
+        }
 
-impl<IN, OUT> fmt::Debug for Callback<IN, OUT> {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "Callback<_>")
-    }
-}
+        impl<IN, OUT> $callback<IN, OUT> {
+            /// This method calls the callback's function.
+            pub fn emit(&self, value: $in_ty) -> OUT {
+                (*self.cb)(value)
+            }
+        }
 
-impl<IN, OUT> Callback<IN, OUT> {
-    /// This method calls the callback's function.
-    pub fn emit(&self, value: IN) -> OUT {
-        (*self.cb)(value)
-    }
+        impl<IN> $callback<IN> {
+            /// Creates a "no-op" callback which can be used when it is not suitable to use an
+            /// `Option<$callback>`.
+            pub fn noop() -> Self {
+                Self::from(|_: $in_ty| ())
+            }
+        }
```

Comment:
I believe this is actually wrong because creating 2 "noop" callbacks and comparing them for equality will yield false

Reply:
I think this is fine as it is trying to compare the address of 2 functions are not the content of the functions.

Reply:
I should have written a test -_-

The PartialEq implementation uses `Rc::ptr_eq()` so it's actually comparing that it is the same RC but we create one each time we call noop (or Default).

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3e6510283a5655f1166410d77fd62d51

Reply:
I made this PR to address this: #3430